File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {
    private List<List<Integer>> res;
    private List<Integer> track;

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        res = new ArrayList<>();
        track = new ArrayList<>();
        Arrays.sort(nums);
        backtrack(nums, target, 0);
        return res;
    }

    private void backtrack(int[] nums, int target, int start) {
        if (target == 0) {
            res.add(new ArrayList<>(track));
            return;
        }

        for (int i = start; i < nums.length; i++) {
            int cur = nums[i];
            if (cur > target) break;
            track.add(cur);
            backtrack(nums, target - cur, i);
            track.remove(track.size() - 1);
        }
    }
}

File: Data Structures & Algorithms/subsets/submission-1.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.res, self.track = [], []
        self.backtrack(nums, 0)
        return self.res
    
    def backtrack(self, nums, start):
        self.res.append(self.track[:])
        n = len(nums)
        if start > n:
            return
        
        for i in range(start, n):
            self.track.append(nums[i])
            self.backtrack(nums, i + 1)
            self.track.pop()
        